// qlobber/test/common.js
"use strict";

exports.expected_trie = {"a":{"b":{"c":{".":["t1","t20"]},"b":{"c":{".":["t4"]},".":["t14"]},".":["t15"]},"*":{"c":{".":["t2"]},".":["t9"]},"#":{"b":{".":["t3"]},".":["t11"],"#":{".":["t12"]}}},"#":{".":["t5"],"#":{".":["t6"],"#":{".":["t24"]}},"b":{".":["t7"],"#":{".":["t26"]}},"*":{"#":{".":["t22"]}}},"*":{"*":{".":["t8"],"*":{".":["t18"]}},"b":{"c":{".":["t10"]}},"#":{".":["t21"],"#":{".":["t23"]}},".":["t25"]},"b":{"b":{"c":{".":["t13"]}},"c":{".":["t16"]}},"":{".":["t17"]},"vodka":{"martini":{".":["t19"]}}};

exports.expected_visits = [
    { type: 'start_entries' },
    { type: 'entry', i: 0, key: 'a' },
    { type: 'start_entries' },
    { type: 'entry', i: 0, key: 'b' },
    { type: 'start_entries' },
    { type: 'entry', i: 0, key: 'c' },
    { type: 'start_entries' },
    { type: 'entry', i: 0, key: '.' },
    { type: 'start_values' },
    { type: 'value', i: 0, value: 't1' },
    { type: 'value', i: 1, value: 't20' },
    { type: 'end_values' },
    { type: 'end_entries' },
    { type: 'entry', i: 1, key: 'b' },
    { type: 'start_entries' },
    { type: 'entry', i: 0, key: 'c' },
    { type: 'start_entries' },
    { type: 'entry', i: 0, key: '.' },
    { type: 'start_values' },
    { type: 'value', i: 0, value: 't4' },
    { type: 'end_values' },
    { type: 'end_entries' },
    { type: 'entry', i: 1, key: '.' },
    { type: 'start_values' },
    { type: 'value', i: 0, value: 't14' },
    { type: 'end_values' },
    { type: 'end_entries' },
    { type: 'entry', i: 2, key: '.' },
    { type: 'start_values' },
    { type: 'value', i: 0, value: 't15' },
    { type: 'end_values' },
    { type: 'end_entries' },
    { type: 'entry', i: 1, key: '*' },
    { type: 'start_entries' },
    { type: 'entry', i: 0, key: 'c' },
    { type: 'start_entries' },
    { type: 'entry', i: 0, key: '.' },
    { type: 'start_values' },
    { type: 'value', i: 0, value: 't2' },
    { type: 'end_values' },
    { type: 'end_entries' },
    { type: 'entry', i: 1, key: '.' },
    { type: 'start_values' },
    { type: 'value', i: 0, value: 't9' },
    { type: 'end_values' },
    { type: 'end_entries' },
    { type: 'entry', i: 2, key: '#' },
    { type: 'start_entries' },
    { type: 'entry', i: 0, key: 'b' },
    { type: 'start_entries' },
    { type: 'entry', i: 0, key: '.' },
    { type: 'start_values' },
    { type: 'value', i: 0, value: 't3' },
    { type: 'end_values' },
    { type: 'end_entries' },
    { type: 'entry', i: 1, key: '.' },
    { type: 'start_values' },
    { type: 'value', i: 0, value: 't11' },
    { type: 'end_values' },
    { type: 'entry', i: 2, key: '#' },
    { type: 'start_entries' },
    { type: 'entry', i: 0, key: '.' },
    { type: 'start_values' },
    { type: 'value', i: 0, value: 't12' },
    { type: 'end_values' },
    { type: 'end_entries' },
    { type: 'end_entries' },
    { type: 'end_entries' },
    { type: 'entry', i: 1, key: '#' },
    { type: 'start_entries' },
    { type: 'entry', i: 0, key: '.' },
    { type: 'start_values' },
    { type: 'value', i: 0, value: 't5' },
    { type: 'end_values' },
    { type: 'entry', i: 1, key: '#' },
    { type: 'start_entries' },
    { type: 'entry', i: 0, key: '.' },
    { type: 'start_values' },
    { type: 'value', i: 0, value: 't6' },
    { type: 'end_values' },
    { type: 'entry', i: 1, key: '#' },
    { type: 'start_entries' },
    { type: 'entry', i: 0, key: '.' },
    { type: 'start_values' },
    { type: 'value', i: 0, value: 't24' },
    { type: 'end_values' },
    { type: 'end_entries' },
    { type: 'end_entries' },
    { type: 'entry', i: 2, key: 'b' },
    { type: 'start_entries' },
    { type: 'entry', i: 0, key: '.' },
    { type: 'start_values' },
    { type: 'value', i: 0, value: 't7' },
    { type: 'end_values' },
    { type: 'entry', i: 1, key: '#' },
    { type: 'start_entries' },
    { type: 'entry', i: 0, key: '.' },
    { type: 'start_values' },
    { type: 'value', i: 0, value: 't26' },
    { type: 'end_values' },
    { type: 'end_entries' },
    { type: 'end_entries' },
    { type: 'entry', i: 3, key: '*' },
    { type: 'start_entries' },
    { type: 'entry', i: 0, key: '#' },
    { type: 'start_entries' },
    { type: 'entry', i: 0, key: '.' },
    { type: 'start_values' },
    { type: 'value', i: 0, value: 't22' },
    { type: 'end_values' },
    { type: 'end_entries' },
    { type: 'end_entries' },
    { type: 'end_entries' },
    { type: 'entry', i: 2, key: '*' },
    { type: 'start_entries' },
    { type: 'entry', i: 0, key: '*' },
    { type: 'start_entries' },
    { type: 'entry', i: 0, key: '.' },
    { type: 'start_values' },
    { type: 'value', i: 0, value: 't8' },
    { type: 'end_values' },
    { type: 'entry', i: 1, key: '*' },
    { type: 'start_entries' },
    { type: 'entry', i: 0, key: '.' },
    { type: 'start_values' },
    { type: 'value', i: 0, value: 't18' },
    { type: 'end_values' },
    { type: 'end_entries' },
    { type: 'end_entries' },
    { type: 'entry', i: 1, key: 'b' },
    { type: 'start_entries' },
    { type: 'entry', i: 0, key: 'c' },
    { type: 'start_entries' },
    { type: 'entry', i: 0, key: '.' },
    { type: 'start_values' },
    { type: 'value', i: 0, value: 't10' },
    { type: 'end_values' },
    { type: 'end_entries' },
    { type: 'end_entries' },
    { type: 'entry', i: 2, key: '#' },
    { type: 'start_entries' },
    { type: 'entry', i: 0, key: '.' },
    { type: 'start_values' },
    { type: 'value', i: 0, value: 't21' },
    { type: 'end_values' },
    { type: 'entry', i: 1, key: '#' },
    { type: 'start_entries' },
    { type: 'entry', i: 0, key: '.' },
    { type: 'start_values' },
    { type: 'value', i: 0, value: 't23' },
    { type: 'end_values' },
    { type: 'end_entries' },
    { type: 'end_entries' },
    { type: 'entry', i: 3, key: '.' },
    { type: 'start_values' },
    { type: 'value', i: 0, value: 't25' },
    { type: 'end_values' },
    { type: 'end_entries' },
    { type: 'entry', i: 3, key: 'b' },
    { type: 'start_entries' },
    { type: 'entry', i: 0, key: 'b' },
    { type: 'start_entries' },
    { type: 'entry', i: 0, key: 'c' },
    { type: 'start_entries' },
    { type: 'entry', i: 0, key: '.' },
    { type: 'start_values' },
    { type: 'value', i: 0, value: 't13' },
    { type: 'end_values' },
    { type: 'end_entries' },
    { type: 'end_entries' },
    { type: 'entry', i: 1, key: 'c' },
    { type: 'start_entries' },
    { type: 'entry', i: 0, key: '.' },
    { type: 'start_values' },
    { type: 'value', i: 0, value: 't16' },
    { type: 'end_values' },
    { type: 'end_entries' },
    { type: 'end_entries' },
    { type: 'entry', i: 4, key: '' },
    { type: 'start_entries' },
    { type: 'entry', i: 0, key: '.' },
    { type: 'start_values' },
    { type: 'value', i: 0, value: 't17' },
    { type: 'end_values' },
    { type: 'end_entries' },
    { type: 'entry', i: 5, key: 'vodka' },
    { type: 'start_entries' },
    { type: 'entry', i: 0, key: 'martini' },
    { type: 'start_entries' },
    { type: 'entry', i: 0, key: '.' },
    { type: 'start_values' },
    { type: 'value', i: 0, value: 't19' },
    { type: 'end_values' },
    { type: 'end_entries' },
    { type: 'end_entries' },
    { type: 'end_entries' }
];

exports.expected_true_visits = [
    { type: 'start_entries' },
    { type: 'entry', i: 0, key: 'a' },
    { type: 'start_entries' },
    { type: 'entry', i: 0, key: 'b' },
    { type: 'start_entries' },
    { type: 'entry', i: 0, key: 'c' },
    { type: 'start_entries' },
    { type: 'entry', i: 0, key: '.' },
    { type: 'start_values' },
    { type: 'value', i: 0, value: true },
    { type: 'end_values' },
    { type: 'end_entries' },
    { type: 'entry', i: 1, key: 'b' },
    { type: 'start_entries' },
    { type: 'entry', i: 0, key: 'c' },
    { type: 'start_entries' },
    { type: 'entry', i: 0, key: '.' },
    { type: 'start_values' },
    { type: 'value', i: 0, value: true },
    { type: 'end_values' },
    { type: 'end_entries' },
    { type: 'entry', i: 1, key: '.' },
    { type: 'start_values' },
    { type: 'value', i: 0, value: true },
    { type: 'end_values' },
    { type: 'end_entries' },
    { type: 'entry', i: 2, key: '.' },
    { type: 'start_values' },
    { type: 'value', i: 0, value: true },
    { type: 'end_values' },
    { type: 'end_entries' },
    { type: 'entry', i: 1, key: '*' },
    { type: 'start_entries' },
    { type: 'entry', i: 0, key: 'c' },
    { type: 'start_entries' },
    { type: 'entry', i: 0, key: '.' },
    { type: 'start_values' },
    { type: 'value', i: 0, value: true },
    { type: 'end_values' },
    { type: 'end_entries' },
    { type: 'entry', i: 1, key: '.' },
    { type: 'start_values' },
    { type: 'value', i: 0, value: true },
    { type: 'end_values' },
    { type: 'end_entries' },
    { type: 'entry', i: 2, key: '#' },
    { type: 'start_entries' },
    { type: 'entry', i: 0, key: 'b' },
    { type: 'start_entries' },
    { type: 'entry', i: 0, key: '.' },
    { type: 'start_values' },
    { type: 'value', i: 0, value: true },
    { type: 'end_values' },
    { type: 'end_entries' },
    { type: 'entry', i: 1, key: '.' },
    { type: 'start_values' },
    { type: 'value', i: 0, value: true },
    { type: 'end_values' },
    { type: 'entry', i: 2, key: '#' },
    { type: 'start_entries' },
    { type: 'entry', i: 0, key: '.' },
    { type: 'start_values' },
    { type: 'value', i: 0, value: true },
    { type: 'end_values' },
    { type: 'end_entries' },
    { type: 'end_entries' },
    { type: 'end_entries' },
    { type: 'entry', i: 1, key: '#' },
    { type: 'start_entries' },
    { type: 'entry', i: 0, key: '.' },
    { type: 'start_values' },
    { type: 'value', i: 0, value: true },
    { type: 'end_values' },
    { type: 'entry', i: 1, key: '#' },
    { type: 'start_entries' },
    { type: 'entry', i: 0, key: '.' },
    { type: 'start_values' },
    { type: 'value', i: 0, value: true },
    { type: 'end_values' },
    { type: 'entry', i: 1, key: '#' },
    { type: 'start_entries' },
    { type: 'entry', i: 0, key: '.' },
    { type: 'start_values' },
    { type: 'value', i: 0, value: true },
    { type: 'end_values' },
    { type: 'end_entries' },
    { type: 'end_entries' },
    { type: 'entry', i: 2, key: 'b' },
    { type: 'start_entries' },
    { type: 'entry', i: 0, key: '.' },
    { type: 'start_values' },
    { type: 'value', i: 0, value: true },
    { type: 'end_values' },
    { type: 'entry', i: 1, key: '#' },
    { type: 'start_entries' },
    { type: 'entry', i: 0, key: '.' },
    { type: 'start_values' },
    { type: 'value', i: 0, value: true },
    { type: 'end_values' },
    { type: 'end_entries' },
    { type: 'end_entries' },
    { type: 'entry', i: 3, key: '*' },
    { type: 'start_entries' },
    { type: 'entry', i: 0, key: '#' },
    { type: 'start_entries' },
    { type: 'entry', i: 0, key: '.' },
    { type: 'start_values' },
    { type: 'value', i: 0, value: true },
    { type: 'end_values' },
    { type: 'end_entries' },
    { type: 'end_entries' },
    { type: 'end_entries' },
    { type: 'entry', i: 2, key: '*' },
    { type: 'start_entries' },
    { type: 'entry', i: 0, key: '*' },
    { type: 'start_entries' },
    { type: 'entry', i: 0, key: '.' },
    { type: 'start_values' },
    { type: 'value', i: 0, value: true },
    { type: 'end_values' },
    { type: 'entry', i: 1, key: '*' },
    { type: 'start_entries' },
    { type: 'entry', i: 0, key: '.' },
    { type: 'start_values' },
    { type: 'value', i: 0, value: true },
    { type: 'end_values' },
    { type: 'end_entries' },
    { type: 'end_entries' },
    { type: 'entry', i: 1, key: 'b' },
    { type: 'start_entries' },
    { type: 'entry', i: 0, key: 'c' },
    { type: 'start_entries' },
    { type: 'entry', i: 0, key: '.' },
    { type: 'start_values' },
    { type: 'value', i: 0, value: true },
    { type: 'end_values' },
    { type: 'end_entries' },
    { type: 'end_entries' },
    { type: 'entry', i: 2, key: '#' },
    { type: 'start_entries' },
    { type: 'entry', i: 0, key: '.' },
    { type: 'start_values' },
    { type: 'value', i: 0, value: true },
    { type: 'end_values' },
    { type: 'entry', i: 1, key: '#' },
    { type: 'start_entries' },
    { type: 'entry', i: 0, key: '.' },
    { type: 'start_values' },
    { type: 'value', i: 0, value: true },
    { type: 'end_values' },
    { type: 'end_entries' },
    { type: 'end_entries' },
    { type: 'entry', i: 3, key: '.' },
    { type: 'start_values' },
    { type: 'value', i: 0, value: true },
    { type: 'end_values' },
    { type: 'end_entries' },
    { type: 'entry', i: 3, key: 'b' },
    { type: 'start_entries' },
    { type: 'entry', i: 0, key: 'b' },
    { type: 'start_entries' },
    { type: 'entry', i: 0, key: 'c' },
    { type: 'start_entries' },
    { type: 'entry', i: 0, key: '.' },
    { type: 'start_values' },
    { type: 'value', i: 0, value: true },
    { type: 'end_values' },
    { type: 'end_entries' },
    { type: 'end_entries' },
    { type: 'entry', i: 1, key: 'c' },
    { type: 'start_entries' },
    { type: 'entry', i: 0, key: '.' },
    { type: 'start_values' },
    { type: 'value', i: 0, value: true },
    { type: 'end_values' },
    { type: 'end_entries' },
    { type: 'end_entries' },
    { type: 'entry', i: 4, key: '' },
    { type: 'start_entries' },
    { type: 'entry', i: 0, key: '.' },
    { type: 'start_values' },
    { type: 'value', i: 0, value: true },
    { type: 'end_values' },
    { type: 'end_entries' },
    { type: 'entry', i: 5, key: 'vodka' },
    { type: 'start_entries' },
    { type: 'entry', i: 0, key: 'martini' },
    { type: 'start_entries' },
    { type: 'entry', i: 0, key: '.' },
    { type: 'start_values' },
    { type: 'value', i: 0, value: true },
    { type: 'end_values' },
    { type: 'end_entries' },
    { type: 'end_entries' },
    { type: 'end_entries' }
];

if (Array.from === undefined)
{
	Array.from = function (s)
	{
		var a = [];
		s.forEach(function (v)
		{
			a[a.length] = v;
		});
		return a;
	};
}

function get_trie(matcher, t)
{
    t = t || matcher.get_trie();
    var k, r = {};
    for (k of t.keys())
    {
        if (k === '.')
        {
            r[k] = Array.from(t.get(k));
        }
        else
        {
            r[k] = get_trie(matcher, t.get(k));
        }
    }
    return r;
}

exports.get_trie = get_trie;
